services/121-service/src/payments/fsp-integration/safaricom/safaricom.service.ts

```diff
@@ -47,14 +49,28 @@ export class SafaricomService
   public async doTransfer(
     transferData: DoTransferParams,
   ): Promise<DoTransferReturnType> {
+    // Store initial transfer record before transfer because of callback
+    const safaricomTransfer =
+      await this.safaricomTransferRepository.storeSafaricomTransfer(
+        transferData.originatorConversationId,
+        transferData.transactionId,
+      );
+
+    // Do transfer
     await this.safaricomApiService.authenticate();
+    const payload = this.createPayload(transferData);
```

Comment:
Functionality should be encapsulated in SafaricomApiService.

Reply:
Updated!